File: 0643-maximum-average-subarray-i/0643-maximum-average-subarray-i.java
class Solution {
    public double findMaxAverage(int[] nums, int k) {
        // double avg = Integer.MIN_VALUE;

        // int left = 0 ;
        // int right = k - 1;

        // while(right < nums.length)
        // {
        // int sum = 0;
        // for(int i = left ; i <= right ; i++)
        // {
        // sum += nums[i];
        // }
        // avg = Math.max(avg, (double) sum / k);
        // left++;
        // right++;
        // }

        // return avg;
        // --------------------------------------------

        double max = Integer.MIN_VALUE;

        int left = 0;
        int right = k - 1;
        int window = 0;

        for (int i = 0; i < k; i++) {
            window += nums[i];
        }

        while (right != nums.length) {
            max = Math.max(max, (double) window / k);

            right++;

            window -= nums[left];
            left++;

            if (right != nums.length) {
                window += nums[right];
            }

        }
        return max;
    }
}

File: 0643-maximum-average-subarray-i/0643-maximum-average-subarray-i.cpp
class Solution {
public:
    double findMaxAverage(vector<int>& nums, int k) {
        // TLE
        // int start = 0;
        // double sum = 0.0;
        // double max_avg = -DBL_MAX;

        // while(start <= nums.size() - k){
        //     for(int i = start ; i < start + k ; i++){
        //         sum += nums[i];
        //     }
        //     max_avg = max(max_avg, sum / k);
        //     start++;
        //     sum = 0;
        // }

        // return max_avg;
        // <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

        double sum = 0;
        for(int i = 0 ; i < k ; i++){
            sum += nums[i];
        }

        double max_avg = sum;
        for(int i = k ; i < nums.size() ; i++){
            sum = sum - nums[i - k] + nums[i];
            max_avg = max(max_avg, sum);
        }
        
        return max_avg / k;
    }
};